# 03 — Decision Dataset Preparation

## Objective

Build the **modeling dataset** for the Next Best Action solution from historical `campaign_sends`.

Each modeling row represents one historical marketing decision:

> **customer × promoted product × channel × send time**

The model will later learn:

> Given this customer, product, channel and context, what is the probability of conversion?

### Target

- `had_conversion`

### Economic evaluation fields

- `send_cost`
- `conversion_value`

These economic fields are retained for evaluation, but **must not be used as model features** because they are known only after the action is taken.

### Leakage policy

We will exclude post-send outcome fields such as:

- `was_delivered`
- `was_opened`
- `was_clicked`
- `click_count`
- `conversion_date`
- `open_device`
- `failure_reason`

For the MVP, the decision engine will focus on channels with observed positive conversions in the historical data:

- Email
- SMS
- Push

Voice and WhatsApp are kept out of the first modeling dataset because the discovery notebook showed zero recorded conversions for those channels.


## 1. Install dependencies


In [1]:
!pip -q install boto3 pyarrow tqdm


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 140.0/140.0 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.0/16.0 MB 88.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.2/90.2 kB 6.2 MB/s eta 0:00:00


## 2. Imports and S3 configuration

Use the same S3 access pattern as the previous notebooks.

Credentials are requested at runtime and are not stored in the notebook.


In [2]:
import boto3
import pandas as pd
import numpy as np
import getpass

from tqdm.auto import tqdm

AWS_ACCESS_KEY_ID = input("AWS Access Key ID: ")
AWS_SECRET_ACCESS_KEY = getpass.getpass("AWS Secret Access Key: ")

AWS_REGION = "us-east-2"
BUCKET_NAME = "factored-datathon-2026-s3-157725502942-us-east-2-an"

session = boto3.Session(
    aws_access_key_id=AWS_ACCESS_KEY_ID,
    aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
    region_name=AWS_REGION
)

s3 = session.client("s3")

print("S3 connection configured.")


AWS Access Key ID: AKIASJOJMFXPDNU7JH7C
AWS Secret Access Key: ··········
S3 connection configured.


## 3. Locate source files

We will use three source tables:

- `campaign_sends`: individual historical marketing sends
- `customers`: customer context
- `marketing_campaigns`: promoted product and campaign context

`campaign_sends` remains at individual-send grain. We do **not** aggregate it as we did for the baseline analysis.


In [3]:
def list_s3_files(bucket, prefix="data/"):
    paginator = s3.get_paginator("list_objects_v2")
    rows = []

    for page in paginator.paginate(
        Bucket=bucket,
        Prefix=prefix
    ):
        for obj in page.get("Contents", []):
            rows.append({
                "key": obj["Key"],
                "size_mb": obj["Size"] / (1024 * 1024),
                "last_modified": obj["LastModified"]
            })

    return pd.DataFrame(rows)


files_df = list_s3_files(
    BUCKET_NAME,
    prefix="data/"
)

campaign_send_keys = sorted(
    files_df.loc[
        files_df["key"]
        .str.startswith("data/campaign_sends/", na=False)
        &
        files_df["key"]
        .str.endswith((".csv", ".csv.gz"), na=False),
        "key"
    ].tolist()
)

customers_matches = files_df[
    files_df["key"].str.contains(
        r"data/customers",
        case=False,
        na=False
    )
].copy()

campaigns_matches = files_df[
    files_df["key"].str.contains(
        r"data/marketing_campaigns",
        case=False,
        na=False
    )
].copy()

print(f"campaign_sends files: {len(campaign_send_keys):,}")
print(f"customers files: {len(customers_matches):,}")
print(f"marketing_campaigns files: {len(campaigns_matches):,}")

display(customers_matches.head())
display(campaigns_matches.head())


campaign_sends files: 1,083
customers files: 1
marketing_campaigns files: 1


,key,size_mb,last_modified
4375,data/customers.csv,44.724806,2026-09-01 02:38:54+00:00


,key,size_mb,last_modified
5474,data/marketing_campaigns.csv,0.031195,2026-09-01 02:48:51+00:00


## 4. Load customer features

For the first model we will use a compact set of customer attributes.

We intentionally avoid personally identifiable fields such as names, document numbers, email, phone and address.

> Note: the supplied customer table behaves as a snapshot. Some mutable attributes may reflect the latest known value rather than the exact value at historical send time. This is a limitation of the MVP and should be documented in the final solution.


In [4]:
customers_key = customers_matches.iloc[0]["key"]

obj = s3.get_object(
    Bucket=BUCKET_NAME,
    Key=customers_key
)

customer_columns = [
    "customer_id",
    "date_of_birth",
    "gender",
    "country",
    "segment",
    "credit_score",
    "estimated_monthly_income",
    "marital_status",
    "education_level",
    "registration_date",
    "customer_status",
    "accepts_marketing",
]

customers_df = pd.read_csv(
    obj["Body"],
    usecols=lambda c: c in customer_columns
)

customers_df["date_of_birth"] = pd.to_datetime(
    customers_df["date_of_birth"],
    errors="coerce"
)

customers_df["registration_date"] = pd.to_datetime(
    customers_df["registration_date"],
    errors="coerce"
)

print("Customers:", customers_df.shape)
print("\nColumns:")
print(customers_df.columns.tolist())

display(customers_df.head())


Customers: (150000, 12)

Columns:
['customer_id', 'date_of_birth', 'gender', 'country', 'segment', 'credit_score', 'estimated_monthly_income', 'marital_status', 'education_level', 'registration_date', 'customer_status', 'accepts_marketing']


,customer_id,date_of_birth,gender,country,segment,credit_score,estimated_monthly_income,marital_status,education_level,registration_date,customer_status,accepts_marketing
0,CLI-G4X2AMVD62NR,1967-06-18,M,Colombia,Plus,701.0,24678431.94,Married,University,2021-07-30 05:39:39,Active,False
1,CLI-F2DZJYU0POJ9,1976-04-16,M,México,Basic,576.0,NaN,Single,NaN,2022-12-09 13:10:47,Active,True
2,CLI-8WU28O74XKHS,1995-02-09,O,México,Basic,657.0,28495.40,Divorced,NaN,2024-04-06 16:26:17,Active,True
3,CLI-IU5Y26LO84W8,1948-06-05,O,México,Basic,598.0,37284.11,NaN,High School,2024-09-10 17:43:08,Inactive,True
4,CLI-C4Z5FPK4YO51,1985-07-22,M,México,Basic,594.0,25257.93,Married,Elementary,2026-03-03 23:05:30,Active,True


## 5. Load campaign context

The campaign table tells us which product was promoted and provides campaign-level context.

These fields are available before the individual customer outcome and can therefore be used as decision features.


In [5]:
campaigns_key = campaigns_matches.iloc[0]["key"]

obj = s3.get_object(
    Bucket=BUCKET_NAME,
    Key=campaigns_key
)

campaign_columns = [
    "campaign_id",
    "campaign_type",
    "campaign_objective",
    "promoted_product",
    "target_segment",
    "target_country",
    "start_date",
    "end_date",
    "expected_conversion_rate",
]

marketing_campaigns_df = pd.read_csv(
    obj["Body"],
    usecols=lambda c: c in campaign_columns
)

marketing_campaigns_df["start_date"] = pd.to_datetime(
    marketing_campaigns_df["start_date"],
    errors="coerce"
)

marketing_campaigns_df["end_date"] = pd.to_datetime(
    marketing_campaigns_df["end_date"],
    errors="coerce"
)

print("Marketing campaigns:", marketing_campaigns_df.shape)

display(marketing_campaigns_df.head())


Marketing campaigns: (200, 9)


,campaign_id,campaign_type,campaign_objective,promoted_product,target_segment,target_country,start_date,end_date,expected_conversion_rate
0,CMP-8LKPPFGAVR5F,Push,Acquisition,Tarjeta Crédito,Plus,NaN,2024-01-12,2024-01-30,1.86
1,CMP-VKP18A283AMD,Email,Retention,Cuenta Ahorro,NaN,Mexico,2025-07-31,2025-10-16,9.47
2,CMP-E2WKP7HVN55Q,Email,Cross-sell,Cuenta Ahorro,Student,Argentina,2024-09-08,2024-09-22,12.17
3,CMP-G9JXO4VRRCA5,Email,Retention,Inversión,Premium,Colombia,2024-05-09,2024-07-31,9.14
4,CMP-4DZT88K2QTZJ,Email,Cross-sell,Cuenta Corriente,NaN,NaN,2024-04-25,2024-05-22,7.58


## 6. Utility functions


In [6]:
def normalize_bool(series):
    if pd.api.types.is_bool_dtype(series):
        return series.astype("int8")

    if pd.api.types.is_numeric_dtype(series):
        return (
            pd.to_numeric(series, errors="coerce")
            .fillna(0)
            .ne(0)
            .astype("int8")
        )

    normalized = (
        series
        .astype(str)
        .str.strip()
        .str.lower()
    )

    return (
        normalized
        .isin(["true", "1", "yes", "y", "t"])
        .astype("int8")
    )


## 7. Build the individual send-level modeling dataset

For every historical send we retain only information that is available at decision time, plus the target and economic evaluation fields.

### Decision-time fields

- customer
- send timestamp
- channel
- campaign
- promoted product
- customer profile
- campaign context

### Outcome fields retained only for evaluation

- `had_conversion`
- `send_cost`
- `conversion_value`

We deliberately do not include opens, clicks, delivery outcome or conversion metadata as model features.


In [7]:
MODEL_CHANNELS = ["Email", "SMS", "Push"]

send_columns = [
    "send_id",
    "send_date",
    "campaign_id",
    "customer_id",
    "send_channel",
    "had_conversion",
    "conversion_value",
    "send_cost",
]

base_parts = []

for key in tqdm(
    campaign_send_keys,
    desc="Building decision dataset"
):
    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    compression = "gzip" if key.endswith(".gz") else None

    part = pd.read_csv(
        obj["Body"],
        compression=compression,
        usecols=lambda c: c in send_columns
    )

    part["send_date"] = pd.to_datetime(
        part["send_date"],
        errors="coerce"
    )

    part["had_conversion"] = normalize_bool(
        part["had_conversion"]
    )

    part["conversion_value"] = pd.to_numeric(
        part["conversion_value"],
        errors="coerce"
    ).fillna(0)

    part["send_cost"] = pd.to_numeric(
        part["send_cost"],
        errors="coerce"
    ).fillna(0)

    # MVP channels with observed positive conversion signal
    part = part[
        part["send_channel"].isin(MODEL_CHANNELS)
    ].copy()

    # Join customer context
    part = part.merge(
        customers_df,
        on="customer_id",
        how="left",
        validate="many_to_one"
    )

    # Join campaign/product context
    part = part.merge(
        marketing_campaigns_df,
        on="campaign_id",
        how="left",
        validate="many_to_one"
    )

    base_parts.append(part)


decision_dataset = pd.concat(
    base_parts,
    ignore_index=True
)

del base_parts

print("Decision dataset built.")
print("Shape:", decision_dataset.shape)


Building decision dataset:   0%|          | 0/1083 [00:00<?, ?it/s]

Decision dataset built.
Shape: (1343128, 27)


## 8. Create time-aware features

All derived features below can be computed at the moment of the send.

We derive:

- customer age at send time
- customer tenure at send time
- send hour
- day of week
- month
- weekend flag

Using the historical send timestamp instead of today's date avoids temporal leakage in these derived features.


In [8]:
decision_dataset["customer_age"] = (
    (
        decision_dataset["send_date"]
        - decision_dataset["date_of_birth"]
    ).dt.days / 365.25
)

decision_dataset["customer_tenure_days"] = (
    decision_dataset["send_date"]
    - decision_dataset["registration_date"]
).dt.days

decision_dataset["send_hour"] = (
    decision_dataset["send_date"].dt.hour
)

decision_dataset["send_day_of_week"] = (
    decision_dataset["send_date"].dt.dayofweek
)

decision_dataset["send_month"] = (
    decision_dataset["send_date"].dt.month
)

decision_dataset["is_weekend"] = (
    decision_dataset["send_day_of_week"]
    .isin([5, 6])
    .astype("int8")
)

decision_dataset["customer_age"] = (
    decision_dataset["customer_age"]
    .where(decision_dataset["customer_age"].between(18, 100))
)

decision_dataset["customer_tenure_days"] = (
    decision_dataset["customer_tenure_days"]
    .clip(lower=0)
)


## 9. Keep only rows that can support a Next Best Action decision

A row without a promoted product cannot train the model to choose a product.

For the MVP we therefore remove:

- rows with missing send timestamp
- rows with missing customer id
- rows whose campaign has no `promoted_product`

We keep missing values in optional customer attributes; those can be handled later by the modeling pipeline.


In [9]:
rows_before = len(decision_dataset)

decision_dataset = (
    decision_dataset
    .dropna(
        subset=[
            "send_date",
            "customer_id",
            "promoted_product",
            "send_channel",
        ]
    )
    .drop_duplicates(
        subset=["send_id"],
        keep="first"
    )
    .reset_index(drop=True)
)

rows_after = len(decision_dataset)

print(f"Rows before cleaning: {rows_before:,}")
print(f"Rows after cleaning:  {rows_after:,}")
print(f"Rows removed:         {rows_before - rows_after:,}")


Rows before cleaning: 1,343,128
Rows after cleaning:  1,214,030
Rows removed:         129,098


## 10. Validate the modeling dataset

Before training anything, verify:

- date range
- number of customers
- number of products
- number of channels
- overall conversion rate
- join completeness
- class imbalance


In [10]:
print("=" * 60)
print("DECISION DATASET VALIDATION")
print("=" * 60)

print(f"Rows: {len(decision_dataset):,}")
print(f"Unique customers: {decision_dataset['customer_id'].nunique():,}")
print(f"Unique campaigns: {decision_dataset['campaign_id'].nunique():,}")
print(f"Unique products: {decision_dataset['promoted_product'].nunique():,}")
print(f"Channels: {sorted(decision_dataset['send_channel'].dropna().unique())}")

print(
    "Date range:",
    decision_dataset["send_date"].min(),
    "→",
    decision_dataset["send_date"].max()
)

print(
    f"Conversion rate: "
    f"{decision_dataset['had_conversion'].mean():.4%}"
)

print("\nTarget distribution:")
display(
    decision_dataset["had_conversion"]
    .value_counts(dropna=False)
    .rename_axis("had_conversion")
    .reset_index(name="rows")
)

print("\nRows with missing customer country:")
print(
    f"{decision_dataset['country'].isna().sum():,}"
)

print("\nRows with missing promoted product:")
print(
    f"{decision_dataset['promoted_product'].isna().sum():,}"
)


DECISION DATASET VALIDATION
Rows: 1,214,030
Unique customers: 149,952
Unique campaigns: 121
Unique products: 7
Channels: ['Email', 'Push', 'SMS']
Date range: 2023-07-01 06:00:42 → 2026-06-18 05:59:53
Conversion rate: 0.7282%

Target distribution:


,had_conversion,rows
0,0,1205190
1,1,8840



Rows with missing customer country:
0

Rows with missing promoted product:
0


## 11. Inspect conversion signal by candidate action

This is only a sanity check.

The model itself will later operate at customer level, but before training we verify that the cleaned decision dataset still preserves meaningful variation by product and channel.


In [11]:
signal_check = (
    decision_dataset
    .groupby(
        [
            "country",
            "promoted_product",
            "send_channel"
        ],
        dropna=False
    )
    .agg(
        sends=("send_id", "size"),
        conversions=("had_conversion", "sum"),
        conversion_rate=("had_conversion", "mean")
    )
    .reset_index()
)

signal_check["conversion_rate_pct"] = (
    signal_check["conversion_rate"] * 100
)

display(
    signal_check
    .sort_values(
        "conversion_rate_pct",
        ascending=False
    )
    .head(30)
)


,country,promoted_product,send_channel,sends,conversions,conversion_rate,conversion_rate_pct
56,México,Préstamo Personal,SMS,32511,336,0.010335,1.033496
20,Argentina,Tarjeta Crédito,SMS,31565,318,0.010074,1.007445
8,Argentina,Inversión,SMS,7688,77,0.010016,1.001561
16,Argentina,Seguro,Push,4808,48,0.009983,0.998336
2,Argentina,Cuenta Ahorro,SMS,13157,130,0.009881,0.988067
26,Colombia,Cuenta Corriente,SMS,15018,146,0.009722,0.972167
44,México,Cuenta Ahorro,SMS,33240,322,0.009687,0.968712
47,México,Cuenta Corriente,SMS,24638,235,0.009538,0.953811
14,Argentina,Préstamo Personal,SMS,12805,121,0.009449,0.944943
53,México,Préstamo Hipotecario,SMS,5950,56,0.009412,0.941176


## 12. Define feature groups for the propensity model

These lists make the modeling contract explicit.

`send_cost`, `conversion_value` and `had_conversion` are **not** included in the feature list.


In [12]:
TARGET = "had_conversion"

NUMERIC_FEATURES = [
    "credit_score",
    "estimated_monthly_income",
    "expected_conversion_rate",
    "customer_age",
    "customer_tenure_days",
    "send_hour",
    "send_day_of_week",
    "send_month",
    "is_weekend",
]

CATEGORICAL_FEATURES = [
    "country",
    "gender",
    "segment",
    "marital_status",
    "education_level",
    "customer_status",
    "accepts_marketing",
    "campaign_type",
    "campaign_objective",
    "promoted_product",
    "target_segment",
    "target_country",
    "send_channel",
]

MODEL_FEATURES = (
    NUMERIC_FEATURES
    + CATEGORICAL_FEATURES
)

ECONOMIC_EVALUATION_FIELDS = [
    "send_cost",
    "conversion_value",
]

print("Numeric features:", len(NUMERIC_FEATURES))
print("Categorical features:", len(CATEGORICAL_FEATURES))
print("Total model features:", len(MODEL_FEATURES))


Numeric features: 9
Categorical features: 13
Total model features: 22


## 13. Temporal train / validation / test split

For this use case we should not use a random split.

The production question is:

> Can historical behavior predict future conversions?

Therefore the evaluation must respect time order.

For the MVP:

- earliest 70% of dates → train
- next 15% → validation
- latest 15% → test

The split boundaries are determined from the observed send-date distribution.


In [13]:
sorted_dates = (
    decision_dataset["send_date"]
    .sort_values()
)

train_cutoff = sorted_dates.quantile(0.70)
validation_cutoff = sorted_dates.quantile(0.85)

train_df = decision_dataset[
    decision_dataset["send_date"] <= train_cutoff
].copy()

validation_df = decision_dataset[
    (decision_dataset["send_date"] > train_cutoff)
    &
    (decision_dataset["send_date"] <= validation_cutoff)
].copy()

test_df = decision_dataset[
    decision_dataset["send_date"] > validation_cutoff
].copy()

print("Train cutoff:", train_cutoff)
print("Validation cutoff:", validation_cutoff)

print("\nSplit sizes:")
print(f"Train:      {len(train_df):,}")
print(f"Validation: {len(validation_df):,}")
print(f"Test:       {len(test_df):,}")

print("\nConversion rates:")
print(f"Train:      {train_df[TARGET].mean():.4%}")
print(f"Validation: {validation_df[TARGET].mean():.4%}")
print(f"Test:       {test_df[TARGET].mean():.4%}")


Train cutoff: 2025-08-18 11:10:03.300000
Validation cutoff: 2026-01-15 19:12:48.750000128

Split sizes:
Train:      849,821
Validation: 182,104
Test:       182,105

Conversion rates:
Train:      0.7324%
Validation: 0.7457%
Test:       0.6908%


## 14. Save the prepared dataset

Saving the cleaned dataset avoids reprocessing all 1,083 `campaign_sends` files every time the Colab runtime restarts.

The default path below writes to the current Colab runtime.

If you want persistence across runtime resets, mount Google Drive and change `OUTPUT_PATH` to a Drive path.


In [14]:
OUTPUT_PATH = "/content/decision_dataset_base.parquet"

decision_dataset.to_parquet(
    OUTPUT_PATH,
    index=False
)

print(f"Saved to: {OUTPUT_PATH}")
print(
    f"File size: "
    f"{__import__('os').path.getsize(OUTPUT_PATH) / (1024**2):.2f} MB"
)


Saved to: /content/decision_dataset_base.parquet
File size: 80.38 MB


## Next step

With this notebook complete, the next notebook should train the **propensity model**.

The model will estimate:

> `P(conversion | customer, product, channel, context)`

Then the Next Best Action layer can score candidate Product × Channel actions for each customer and rank them using predicted conversion probability plus economic constraints.
